scripts/build-zephyr.sh

```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+if [ "`whoami`" = "root" ]
+then
+	echo "Running the script as root is not permitted"
+	exit 1
+fi
+
+CALLED=$_
+[[ "${BASH_SOURCE[0]}" != "${0}" ]] && SOURCED=1 || SOURCED=0
+
+SCRIPT_SRC=$(realpath ${BASH_SOURCE[0]})
+SCRIPT_DIR=$(dirname $SCRIPT_SRC)
+TOP_DIR=$(realpath $SCRIPT_DIR/..)
+
+if [ $SOURCED = 1 ]; then
+	echo "You must run this script, rather then try to source it."
+	echo "$SCRIPT_SRC"
+	exit 1
+fi
+
+if [ -z "$HDMI2USB_ENV" ]; then
+	echo "You appear to not be inside the HDMI2USB environment."
+	echo "Please enter environment with:"
+	echo "  source scripts/enter-env.sh"
+	exit 1
+fi
+
+if [ -z "$ZEPHYR_SDK_INSTALL_DIR" ]; then
+	echo "Zephyr SDK not found"
+	echo "Install Zephyr SDK and/or set ZEPHYR_SDK_INSTALL_DIR variable."
+	echo "For details, see: https://www.zephyrproject.org/developers/"
+	exit 1
+fi
+
+# Imports TARGET, PLATFORM, CPU and TARGET_BUILD_DIR from Makefile
+eval $(make env)
+make info
+
+set -x
+set -e
+
+ZEPHYR_REPO=https://github.com/zephyrproject-rtos/zephyr
+ZEPHYR_REPO_BRANCH=master
+
+case $CPU in
+	vexriscv)
+		TARGET_BOARD=arty_litex_vexriscv
```

Comment:
This shouldn't be set? Zephyr should run on any board which supports litex right?

Reply:
I got it running on minerva, in renode. Haven't tested on picorv32.

Reply:
PicoRV32 is a bit different as it has non-standard interrupt handling and you need a custom Zephyr port for it.

Reply:
> This shouldn't be set? Zephyr should run on any board which supports litex right?

I understand you mean `arty` part, right?

You are right that it would be much nicer to have just `TARGET_BOARD=litex_vexriscv`, but `arty` dependency is needed as it provides the amount of RAM.

I think it is a good idea to rename the board so that `arty` part is not so prominent (to sth like `litex_vexriscv_arty`), but I'm not sure if we can get rid of it completely.

Reply:
You can't hardcode the amount of memory based on the target. You need to get the amount of memory from the litex generated memory map. See [regions.ld](https://github.com/timvideos/HDMI2USB-firmware-prebuilt/blob/master/archive/master/v0.0.4-455-gde59df1/arty/base/vexriscv/software/include/generated/regions.ld) and [mem.h](https://github.com/timvideos/HDMI2USB-firmware-prebuilt/blob/master/archive/master/v0.0.4-455-gde59df1/arty/base/vexriscv/software/include/generated/mem.h).

If there needs to be a certain amount of RAM, we should be checking that.

Reply:
In general you are right, LiteX (even assuming that we simplify it to one type of core, VexRiscv in this case) is dynamic and highly configurable so there is no *one* LiteX+VexRiscv platform.

We need to provide *some* default configuration of LiteX, though, so it's possible to build Zephyr without building LiteX itself.
Recreating LiteX defaults seems to be a sane choice and this is where `arty_litex_vexriscv` came from.

What we are missing currently is the ability to **reconfigure** the platform dynamically.
This is something we have to address before merging the PR to Zephyr.
The options I currently see are:

* conditionally include `generated/mem.h` / `generated/csr.h` from LiteX in Zephyr's `board/riscv32/litex_vexriscv/board.h` and redefine there values read from the default `dts` (the one for LiteX+VexRiscv on Arty),

* conditionally include `generated/mem.h` / `generated/csr.h` from LiteX in the `dts` itself and use ifdefs and defines directly there,

* add a conditional include of custom `dtsi` file at the end of the default `dts` and pass a dynamic `dtsi` generated by `build-zephyr.sh` script (that would, of course, use data from headers generated by LiteX).

Reply:
We will proceed with this shortly, but in the meantime I believe it would be better to merge https://github.com/zephyrproject-rtos/zephyr/pull/14915 to master first as it has already been reviewed. Then we could issue another PR to Zephyr with a code making the LiteX platform more flexible. What do you think about it?

Reply:
@mateusz-holenko: I'm not well aware of the efforts needed to get a PR merged in Zephyr so if it's easier for you, you could do this step by step and indeed address the genericity in another PR. We could try to help on that too.

Reply:
I think either of your last top options make sense.

BTW I started (but never finished) a tool to take a `csr.cvs` file and generate a [Linux Device Tree file](https://github.com/mithro/litex-devicetree) - see https://github.com/mithro/litex-devicetree